### PR TITLE
CacheFilter: parses the allowed_vary_headers from the cache config.

### LINKED ISSUE #10131: CacheFilter: Parse and handle 'vary' response headers
CacheFilter doesn't yet notice 'vary' response headers. At a minimum, it should refuse to cache such responses. For full functionality, it should enforce CacheConfig.allowed_vary_headers during insertion, and copy listed headers to LookupRequest::vary_headers_ during lookup.

Doing at least the minimum is a blocker for deployment.

### PR DESCRIPTION
**Pull request by cbdm**
Commit Message: CacheFilter: parses the allowed_vary_headers from the cache config.

Signed-off-by: Caio caiomelo@google.com

Additional Description:
Parses the allowlist from the cache config proto; this allows users to define a set of rules to control which headers can be varied in the cache.

Risk Level: Low
Testing: Unit testing
Docs Changes: Updated cache proto's comments regarding `allowed_vary_headers`
Release Notes: N/A
Fixes #10131

### CODE REVIEW

**Comment by jmarantz** [comment]
Can you make a C++ class for the allow-list rather than passing around a vector of matchers?

Also is this a good time to introduce a config object so we don't have to reconstruct the allow structure on every request?

**Comment by cbdm** [comment]
Could you point me to a configuration example in an existing filter? I'm not sure how to proceed with that one.

**Comment by jmarantz** [comment]
https://github.com/envoyproxy/envoy/blob/master/source/extensions/filters/http/gzip/gzip_filter.h

**Comment by cbdm** [comment]
@jmarantz thanks for the link! I'll look into how we can change the config of the plugin to be closer to gzip.

File: source/extensions/filters/http/cache/cache_headers_utils.h [hunk 1]
```diff
@@ -114,10 +114,6 @@ class CacheHeadersUtils {
 
 class VaryHeader {
 public:
-  // Checks if the headers contain an allowed value in the Vary header.
-  static bool isAllowed(const std::vector<Matchers::StringMatcherPtr>& allowlist,
-                        const Http::ResponseHeaderMap& headers);
-
   // Checks if the headers contain a non-empty value in the Vary header.
   static bool hasVary(const Http::ResponseHeaderMap& headers);
```
Comment (jmarantz) [inline source/extensions/filters/http/cache/cache_headers_utils.h:118]: Should these static methods still be public? Or will they now be used only internal to the class as helpers?

File: source/extensions/filters/http/cache/cache_headers_utils.h [hunk 2]
```diff
@@ -128,16 +124,24 @@ class VaryHeader {
   // Parses the header names that are in the Vary header value.
   static std::vector<std::string> parseHeaderValue(const Http::HeaderEntry* vary_header);
 
+  // Parses the allowlist from the Cache Config into the object's private allowlist_.
+  VaryHeader(const Protobuf::RepeatedPtrField<envoy::type::matcher::v3::StringMatcher>& allowlist);
```
Comment (jmarantz) [inline source/extensions/filters/http/cache/cache_headers_utils.h:128]: nit: s/allowlist/allow_list/g

File: source/extensions/filters/http/cache/cache_headers_utils.h [hunk 3]
```diff
@@ -128,16 +124,24 @@ class VaryHeader {
...
   // request.
-  static Http::RequestHeaderMapPtr
-  possibleVariedHeaders(const std::vector<Matchers::StringMatcherPtr>& allowlist,
-                        const Http::RequestHeaderMap& request_headers);
-
+  Http::RequestHeaderMapPtr
+  possibleVariedHeaders(const Http::RequestHeaderMap& request_headers) const;
   // Parses the allowlist of matching rules that define header names that can be used to create
   // varied responses.
   static std::vector<Matchers::StringMatcherPtr> parseAllowlist(
```
Comment (jmarantz) [inline source/extensions/filters/http/cache/cache_headers_utils.h:139]: ditto -- can this static be made private? or just inlined into the constructor?

**Review by jmarantz — COMMENTED** [review]
sorry for the delay -- I thought you might have been planning to do the conversion to the explicit config as part of this PR.

Mostly looks good; just some nits. Can you leave a TODO at least if you are not going to do this conversion to an explicit config in this PR?

/wait

File: source/extensions/filters/http/cache/cache_headers_utils.h (hunk 1 above)
Comment (cbdm) [reply source/extensions/filters/http/cache/cache_headers_utils.h:118]: `hasVary` and `createVaryKey` are used by `simple_http_cache`, so those should stay public. I'll make the other two methods private.

File: source/extensions/filters/http/cache/cache_headers_utils.h (hunk 2 above)
Comment (cbdm) [reply source/extensions/filters/http/cache/cache_headers_utils.h:128]: I'm not familiar with the `/g` here, could you explain what it means?

File: source/extensions/filters/http/cache/cache_headers_utils.h (hunk 3 above)
Comment (cbdm) [reply source/extensions/filters/http/cache/cache_headers_utils.h:139]: Yes, this one can be private (or restructured) as it's only used by the constructor.

**Comment by cbdm** [comment]
Sorry for the delay on my end too; I'm starting to look at the config just now because of the long weekend.
We do have a TODO added from the previous PR, I could expand it to say the gzip is a good example to mirror.

File: source/extensions/filters/http/cache/cache_filter.h [hunk 4]
```diff
@@ -88,8 +88,8 @@ class CacheFilter : public Http::PassThroughFilter,
 
   // TODO(#12901): The allowlist could be constructed only once directly from the config, instead of
```
Comment (cbdm) [inline source/extensions/filters/http/cache/cache_filter.h:89]: @jmarantz there's this TODO that reference the open issue; should I expand on this to mention the gzip filter?

File: source/extensions/filters/http/cache/cache_headers_utils.h (hunk 2 above)
Comment (jmarantz) [reply source/extensions/filters/http/cache/cache_headers_utils.h:128]: it is 'sed' syntax borrowed for code reviews; means do this globally.

File: source/extensions/filters/http/cache/cache_filter.h (hunk 4 above)
Comment (jmarantz) [reply source/extensions/filters/http/cache/cache_filter.h:89]: thanks; works for me.

File: source/extensions/filters/http/cache/cache_headers_utils.h (hunk 3 above)
Comment (cbdm) [reply source/extensions/filters/http/cache/cache_headers_utils.h]: I moved it inside the constructor.

File: source/extensions/filters/http/cache/cache_headers_utils.h (hunk 1 above)
Comment (cbdm) [reply source/extensions/filters/http/cache/cache_headers_utils.h:122]: I moved the logic of `parseAllowlist` to the constructor and made `parseHeaderValues` more generic and moved it into `CacheHeaderUtils`.
Let me know if that's okay.

**Comment by cbdm** [comment]
Hi, from the discussion in #12901, I feel like the explicit config should be a separate PR. What do you think?

**Review by jmarantz — DISMISSED** [review]
master merge needed

**Comment by cbdm** [comment]
I was already working on it :-)

**Comment by cbdm** [comment]
@jmarantz gentle ping after merge and all tests passed.

**Review by jmarantz — APPROVED** [review]

**Pull request merged**